cms/djangoapps/contentstore/utils.py

```diff
@@ -2394,3 +2394,43 @@ def create_or_update_xblock_upstream_link(xblock, course_key: str | CourseKey, c
         version_declined=xblock.upstream_version_declined,
         created=created,
     )
+
+
+def _create_or_update_container_link(course_key: CourseKey, created: datetime | None, xblock):
+    """
+    Create or update upstream->downstream link for containers in database for given xblock.
+    """
+    upstream_container_key = LibraryContainerLocator.from_string(xblock.upstream)
+    try:
+        lib_component = get_container_from_key(upstream_container_key)
+    except ObjectDoesNotExist:
+        log.error(f"Library component not found for {upstream_container_key}")
+        lib_component = None
+    ContainerLink.update_or_create(
+        lib_component,
+        upstream_container_key=upstream_container_key,
+        upstream_context_key=str(upstream_container_key.context_key),
+        downstream_context_key=course_key,
+        downstream_usage_key=xblock.usage_key,
+        version_synced=xblock.upstream_version,
+        version_declined=xblock.upstream_version_declined,
+        created=created,
+    )
+
+
+def create_or_update_xblock_upstream_link(xblock, course_key: CourseKey, created: datetime | None = None) -> None:
+    """
+    Create or update upstream->downstream link in database for given xblock.
+    """
+    if not xblock.upstream:
+        return None
+    try:
+        # Try to create component link
+        _create_or_update_component_link(course_key, created, xblock)
+    except InvalidKeyError:
+        # It is possible that the upstream is a container and UsageKeyV2 parse failed
+        # Create upstream container link
+        try:
+            _create_or_update_container_link(course_key, created, xblock)
+        except InvalidKeyError:
+            log.error(f"Invalid key: {xblock.upstream}")
```

Comment:
Nit for later follow-up: It seems bad to let this error pass with just error logging, since it likely means something more basic is broken upstream.